### PR TITLE
build: Support Mounted Resource Volumes

### PR DESCRIPTION
**Pull request by adambkaplan**
Proposal to support mounts for Secrets and ConfigMaps in builds. This
augments prior work which let Secrets and ConfigMaps be used as sources
in builds. Unlike the current approach, this enhancement leverages
buildah's volume mount feature to let content be available only at build
time, and not at runtime. This is useful for builds that need to access
private artifact repositories or RHEL subscription content.

This current proposal builds on top of Ben Parees's original volume
mount proposal. Supported volume types will be gated at the API level to
ensure the volume type does not pose a security risk and is correctly
lifecycled. To distinguish Kubernetes volumes+mounts from buildah's
volume mount mechanism, the terms "volume content," "input volume,"
"buildah volume mount," and "buildah runtime environment" were
introduced. The proposed API uses these terms to distinguish build
volume mounts from pod volume mounts. Documentation requirements for
this feature were added in the "Drawbacks" section of the proposal.

Because the build controller uses the privileged security context
constraint, build pods to bypass most security features in OpenShift.
Therefore, future volume mounts in build pods need to be tightly
controlled. Open-ended volume types like CSI volumes could be an attack
vector if a developer uses an insecure CSI driver implementation. This
proposal establishes graduation criteria for adding new volume source
types to builds, including security, testing, lifecycle concerns,
failure modes, and feature gating.

### CODE REVIEW

File: enhancements/builds/volume-mounted-resources.md
Comment (bparees) [inline enhancements/builds/volume-mounted-resources.md:89]: and CSI volumes

**Review by bparees — COMMENTED** [review]
one nit, otherwise looks as good as the day i wrote it :)

File: enhancements/builds/volume-mounted-resources.md
Comment (bparees) [inline enhancements/builds/volume-mounted-resources.md:153]: doing this, instead of reusing the k8s volume type and adding validation/constraints that control which volumetypes are allowed may be a sticking point for people like @deads2k or @smarterclayton 

personally i'm ok with it (though it does carry a slightly higher maintenance burden for you), but i'd like to hear from them just in case they are fundamentally opposed.

Comment (adambkaplan) [reply enhancements/builds/volume-mounted-resources.md:153]: I decided to go this route because the real meat of the k8s Volume API is in the `VolumeSource` object, which to my understanding is a bag of pointers. We get the added maintenance burden of validating one and only one source type is set, which we don't _have_ to do on admission (though this is probably a good idea).

Comment (adambkaplan) [reply enhancements/builds/volume-mounted-resources.md:153]: On the flip side, we benefit by gating volume types in the API - invalid volume types fail fast.

File: enhancements/builds/volume-mounted-resources.md
Comment (gabemontero) [inline enhancements/builds/volume-mounted-resources.md:66]: should some reference to @bparees 's original volume mount proposal that you noted in the PR description be made here?

I'm assuming that is at least part of the future being referenced here

I'm also assuming there is no superceding or replacement of @bparees 's original EPs in this space.  Just some "augmentation"

But some brief text describing the relationships, overlapping, non-overlapping elements, even some sort of roadmap-ish flow seems useful and possibly even necessary here

File: enhancements/builds/volume-mounted-resources.md
Comment (gabemontero) [inline enhancements/builds/volume-mounted-resources.md:191]: At least with the current projected resource / share / csi implementation we do not support read only because we have to copy the data from memory (the controller cache) to the mount location after the mount happens... we are not mounting from one existing fs to another like can occur in the more general volume cases

And this is on initial creation, not just updating later on if the secret or configmap changes and the controller receives an event for it.

So minimally this godoc, and the progression of things in general, would have to note that caveat for our new driver

Or we talk about further features and complications to https://github.com/openshift/csi-driver-projected-resource to manage multiple tmpfs  ... one we can write to / update and the one that is used for the volume the driver returns back to the kubelet

File: enhancements/builds/volume-mounted-resources.md
Comment (gabemontero) [inline enhancements/builds/volume-mounted-resources.md:194]: So I almost posed this question above, but chose to do it here with the literal struct golang definition.

I'll copy the text from above for reference:

```
These fields will be wired, via the build controller, directly to the build pod that is constructed.
In addition all mounts into the pod will be done at a path of our choosing, not the VolumeMount path specified, to ensure the user cannot overwrite critical function inside the build pod and use it as an escalation pathway.
Example - all volumes are mounted under `/var/run/openshift.io/volumes` in the build pod containers.
```

So I'm trying to reconcile those words above, including `mounts into the pod will be done at a path of our choosing, not the VolumeMount path specified` with this presence of this field.

Is this a relative path off of `/var/run/openshift.io/volumes` then?  Or are we talking pod file system vs. build container context blah blah blah sort of thing, and this means transient mounts when we call buildah from openshift/builder ?

If so, use of `relative path` and `build context`  in the godoc above seems needed to me to make this clear.

If it is something other then relative path, I need more explanation in general.

File: enhancements/builds/volume-mounted-resources.md
Comment (gabemontero) [inline enhancements/builds/volume-mounted-resources.md:234]: yeah these don't seem like relative paths .... or again maybe it is a Pod FS vs. build context sort of thing

File: enhancements/builds/volume-mounted-resources.md
Comment (gabemontero) [inline enhancements/builds/volume-mounted-resources.md:267]: Yep until the move into the CVO via the storage operators happens, once openshift/api and openshift/client-go are updated with the new fields, we can vendor them into https://github.com/openshift/csi-driver-projected-resource and build upon the existing e2e's there to create BC's that use the new stuff.

**Review by gabemontero — CHANGES_REQUESTED** [review]
Took a first pass at this @bparees @adambkaplan 

Ranges from some clarification on verbiage in godoc and proposal doc ... though I won't argue too too much if you all think I'm being overly precise.

The read only bit and some of the internal details of how the projected resource driver currently work probably will engender more discussion and possibly more changes to the driver at some point.

File: enhancements/builds/volume-mounted-resources.md
Comment (bparees) [reply enhancements/builds/volume-mounted-resources.md:194]: >  Or are we talking pod file system vs. build container context blah blah blah sort of thing, and this means transient mounts when we call buildah from openshift/builder ?

yeah, it's making that distinction.  We will choose the mount location within the build pod.  The user will choose the mount location within the buildah spawned container (the one that's executing the Dockerfile commands).

File: enhancements/builds/volume-mounted-resources.md
Comment (sttts) [reply enhancements/builds/volume-mounted-resources.md:152]: "validating one and only one source type is set" can be done purely in OpenAPI, no need for admission.

File: enhancements/builds/volume-mounted-resources.md
Comment (sttts) [inline enhancements/builds/volume-mounted-resources.md:179]: OpenAPI to restrict to exactly one:
```
oneOf:
- required: ["secret"]
- required: ["configMap"]
- required: ["csi"]
```
This can be added to the CRD OpenAPI schema using a yaml patch (kubebuilder has no marker for that).

File: enhancements/builds/volume-mounted-resources.md
Comment (deads2k) [inline enhancements/builds/volume-mounted-resources.md:120]: don't use object reference.  Create your own type to match.  See doc on https://github.com/kubernetes/api/blob/master/core/v1/types.go#L5366-L5381

File: enhancements/builds/volume-mounted-resources.md
Comment (deads2k) [inline enhancements/builds/volume-mounted-resources.md:134]: will this ever be defaulted to true?  Basically we learned over time that nothing is a bool and it's better to have an enum with something like `MakeStrategy` with values `{Full, Incremental}`

File: enhancements/builds/volume-mounted-resources.md
Comment (deads2k) [inline enhancements/builds/volume-mounted-resources.md:137]: Is this really an `ImagePrePullPolicy` with values `Always`, `IfNotPresent`, and `Never` like on pods?

File: enhancements/builds/volume-mounted-resources.md
Comment (deads2k) [inline enhancements/builds/volume-mounted-resources.md:125]: Same comment about reusing the object reference types.

File: enhancements/builds/volume-mounted-resources.md
Comment (deads2k) [reply enhancements/builds/volume-mounted-resources.md:152]: > doing this, instead of reusing the k8s volume type and adding validation/constraints that control which volumetypes are allowed may be a sticking point for people like @deads2k or @smarterclayton

Having the local help and the constrained choice and avoiding automatic additions are all good properties you get with this approach.

File: enhancements/builds/volume-mounted-resources.md
Comment (deads2k) [inline enhancements/builds/volume-mounted-resources.md:167]: This looks like a discriminated union.  In order to add somethign safely in the future with an old controller running, you need a `VolumeType string` field.  That way the old controller can read the volumetype and realize he doesn't support the one chosen.

File: enhancements/builds/volume-mounted-resources.md
Comment (deads2k) [inline enhancements/builds/volume-mounted-resources.md:171]: Be sure you really intend to keep up with this. Someone added specific keys to path mappings.  Will you support them?  If so, this is fine.  If this is a blind pass through, that could be ok as long as you're sure you don't accidentally change the mounting rules for users.

File: enhancements/builds/volume-mounted-resources.md
Comment (deads2k) [inline enhancements/builds/volume-mounted-resources.md:179]: How will this be controlled with respect to SCCs allowed for users?  I'm actually unsure today how CSI volumes are controlled via SCC

File: enhancements/builds/volume-mounted-resources.md
Comment (deads2k) [inline enhancements/builds/volume-mounted-resources.md:185]: For this struct, please get a review from the storage team.  I'm not as intimately familiar with volume mount problems like path traversals and readonly as they are.

File: enhancements/builds/volume-mounted-resources.md
Comment (adambkaplan) [reply enhancements/builds/volume-mounted-resources.md:191]: It is worth clarifying that this is the mount from the build container into buildah's sub-container. From that perspective we don't need to worry if the CSI driver does a writable volume mount, we can tell buildah to mount a directory into the build sub-container as read-only.

File: enhancements/builds/volume-mounted-resources.md
Comment (adambkaplan) [reply enhancements/builds/volume-mounted-resources.md:66]: This is a true co-authoriship - a big chunk of the text is Ben's original proposal. That never merged, I linked the two via GitHub comments.

File: enhancements/builds/volume-mounted-resources.md
Comment (adambkaplan) [reply enhancements/builds/volume-mounted-resources.md:120]: This is our existing API, I don't think we can change it without breaking things.

File: enhancements/builds/volume-mounted-resources.md
Comment (adambkaplan) [reply enhancements/builds/volume-mounted-resources.md:134]: Existing API, can't change without breaking users.

**Comment by deads2k** [comment]
I like the way this API mirrors volumes.  I want to be sure

1. we don't accidentally escalate users by adding this feature - this is the dev-ex team's responsibility
2. that SCC can actually handle CSI volumes in some reasonable way - this the apiserver or storage team's responsibility

File: enhancements/builds/volume-mounted-resources.md
Comment (adambkaplan) [reply enhancements/builds/volume-mounted-resources.md]: removed CSI volumes :(

File: enhancements/builds/volume-mounted-resources.md
Comment (adambkaplan) [reply enhancements/builds/volume-mounted-resources.md:200]: updated API docs and proposal to distinguish "pod volume mount" from "buildah bind mount"

File: enhancements/builds/volume-mounted-resources.md
Comment (adambkaplan) [reply enhancements/builds/volume-mounted-resources.md:203]: Proposal and API doc updated to distinguish "pod volume mount" from "buildah volume mount"

File: enhancements/builds/volume-mounted-resources.md
Comment (adambkaplan) [reply enhancements/builds/volume-mounted-resources.md]: Existing API in Builds

File: enhancements/builds/volume-mounted-resources.md
Comment (adambkaplan) [reply enhancements/builds/volume-mounted-resources.md]: Existing API in builds

File: enhancements/builds/volume-mounted-resources.md
Comment (adambkaplan) [reply enhancements/builds/volume-mounted-resources.md:174]: Added the discriminator field. My understanding is that we should set/validate this field on admission, in the same way we do with the build strategy type discriminator. Is this correct?

File: enhancements/builds/volume-mounted-resources.md
Comment (adambkaplan) [reply enhancements/builds/volume-mounted-resources.md]: Due to concerns with SCCs, CSI volume sources were dropped from this EP. The build controller uses the privileged SCC to do "safe" host path mounts in build pods, and even with builds in user namespaces we will likely keep these host path mounts.

CSI volume sources can be reintroduced in the future. I think we can gate the allowed CSI driver for these volume types in one of two ways (perhaps not mutually exclusive):

- Enhance SCCs to allowlist drivers for `CSI` volume sources.
- Enhance the cluster Build controller config to allowlist drivers for CSI volume mounts.

File: enhancements/builds/volume-mounted-resources.md
Comment (adambkaplan) [inline enhancements/builds/volume-mounted-resources.md:194]: /cc @nalind 

For mount options supported by buildah.

File: enhancements/builds/volume-mounted-resources.md
Comment (adambkaplan) [inline enhancements/builds/volume-mounted-resources.md:313]: /cc @mrunalp 

I believe we use `mounts.conf` to get RHEL entitlements into workloads from the host, which is relevant for a) RHEL7 nodes, and b) clusters that use MacineConfig to add RHEL entitlements to nodes.

In both situations, these are work-arounds for our lack of entitlements support in RHCOS.

**Comment by adambkaplan** [comment]
> I want to be sure
> 
> 1. we don't accidentally escalate users by adding this feature - this is the dev-ex team's responsibility
> 2. that SCC can actually handle CSI volumes in some reasonable way - this the apiserver or storage team's responsibility

Based on our conversation in slack and my own investigation, we need additional build controller configuration to gate CSI volume mounts. I think this is worth pursuing in a separate enhancement proposal, since ephemeral CSI drivers will be a net-new thing for OpenShift.

Enhancing SCCs probably needs to be added to the original EP for the projected resource CSI driver. We can only allow an SCC to allow all or no CSI volume mounts, which is not sufficient protection.

File: enhancements/builds/volume-mounted-resources.md
Comment (bparees) [reply enhancements/builds/volume-mounted-resources.md]: so if CSI volumes are being dropped from the enhancement, how does this move us towards the shared entitlement cred use case?

(don't get me wrong, it's still a useful evolution of the api and configmap/secret volume mounting has existing use cases/demand)

Comment (adambkaplan) [reply enhancements/builds/volume-mounted-resources.md]: We don't get the benefit of shared entitlements, but we remove developer pain with the experience of adding secrets. Especially for the Docker strategy, where developers are on their own to remove the secret and squash layers.

Adding CSI volumes need to be added in a follow-up EP, where we will need to carefully consider how we secure builds with CSI volumes.

Comment (adambkaplan) [reply enhancements/builds/volume-mounted-resources.md]: (note - we still get the foundation of an API to add volumes. From a coding perspective this will be 80% of what we need to do to support the shared entitlement use case).

Comment (bparees) [reply enhancements/builds/volume-mounted-resources.md]: got it.

File: enhancements/builds/volume-mounted-resources.md
Comment (nalind) [reply enhancements/builds/volume-mounted-resources.md:197]: The library accepts options for controlling propagation ("shared", "rshared", "private", "rprivate", "slave", "rslave"), regular options ("rw, "ro", "dev"/"nodev", "suid"/"nosuid", "exec"/"noexec"), and SELinux labeling controls that I think end up being ignored since we're already in a container ("Z", "z").
There are also a couple of custom ones: "U" to have the ownership of the volume's contents changed to match the current USER value, and "O" which causes the volume source to be used as the lower layer for an overlay mount at the mount point, using a per-run temporary directory as the upper (write) layer.  I don't know that either of those is particularly useful in builds, and of course their functionality willl be limited in the future by the permissions that the builder binary has.

File: enhancements/builds/volume-mounted-resources.md
Comment (nalind) [inline enhancements/builds/volume-mounted-resources.md:277]: Currently, when multiple volumes are set to be mounted at a given location, we choose one of them and proceed.  This will require some patching.

Comment (adambkaplan) [reply enhancements/builds/volume-mounted-resources.md:277]: I think this is something we can do in openshift/builder or in the build controller. Buildah itself doesn't have to fail here (unless you believe this is buggy behavior).

File: enhancements/builds/volume-mounted-resources.md
Comment (coreydaley) [inline enhancements/builds/volume-mounted-resources.md:182]: @adambkaplan Do we really need this since we can infer the type from whether or not `Source` or `ConfigMap` are nil? Then we can just kick it back if both are not nil? (or maybe use a validation?)

Comment (adambkaplan) [reply enhancements/builds/volume-mounted-resources.md:182]: Yes - this is a type discriminator that helps us address controller version skew. As we add more supported volume types, we will run into a situation on upgrade where the apiserver will support new volume types, but the controller won't. This field will need to be set by openshift-apiserver if it is empty, in the same way we set the strategy type.

File: enhancements/builds/volume-mounted-resources.md
Comment (mrunalp) [reply enhancements/builds/volume-mounted-resources.md:316]: @adambkaplan Today we don't allow cluster admins to modify the mounts.conf through ContaineRuntineConfig CRD. They can of course directly modify any such file using the MCO but that is unsupported.

File: enhancements/builds/volume-mounted-resources.md
Comment (adambkaplan) [reply enhancements/builds/volume-mounted-resources.md:197]: Good to know - we will start with ReadOnly `ro` as the only supported option.

**Comment by adambkaplan** [comment]
@bparees @deads2k @sttts @mrunalp all comments have been addressed.

Final call for review - otherwise I think this is good to merge.

File: enhancements/builds/volume-mounted-resources.md
Comment (bparees) [inline enhancements/builds/volume-mounted-resources.md:241]: this is no longer valid/applicable since csi volume types aren't supported.

File: enhancements/builds/volume-mounted-resources.md
Comment (bparees) [inline enhancements/builds/volume-mounted-resources.md:248]: remove this as well

File: enhancements/builds/volume-mounted-resources.md
Comment (bparees) [inline enhancements/builds/volume-mounted-resources.md:276]: we should be able to generate our own volume names(both on the pod and in buildah) such that we avoid collisions.  We don't have to use the volume name the user provides, when setting up volumes.

File: enhancements/builds/volume-mounted-resources.md
Comment (bparees) [inline enhancements/builds/volume-mounted-resources.md:285]: is there a risk that a user can create a build that mounts a secret or configmap that they themselves don't have permission to view?

i think we mitigate that in the existing secret injection logic, though i don't recall how.  We need to ensure it is mitigated here as well.

**Review by bparees — CHANGES_REQUESTED** [review]

File: enhancements/builds/volume-mounted-resources.md
Comment (adambkaplan) [reply enhancements/builds/volume-mounted-resources.md:285]: I don't see any such logic in the build controller. Secrets and ConfigMap volume mounts are local object references in the pod spec. My understanding is that if a Secret exists in a namespace, it can be used by any service account in said namespace.

Comment (bparees) [reply enhancements/builds/volume-mounted-resources.md:285]: i guess it is constrained by "you can't create builds if you can't create pods"

File: enhancements/builds/volume-mounted-resources.md
Comment (adambkaplan) [reply enhancements/builds/volume-mounted-resources.md:276]: True about creating our own volume names, though we get into fun situations where we need to validate/truncate superficially long volume names.

Comment (adambkaplan) [reply enhancements/builds/volume-mounted-resources.md:269]: @bparees I made this a "we can change this in a future enhancement" given the short window we have to land this in 4.8

Comment (bparees) [reply enhancements/builds/volume-mounted-resources.md:269]: btw i wasn't suggesting random suffixes.  you don't have to use any part of the user given name, you can just call it "vol1", "vol2", etc.  So no issue w/ truncation.  Again we already do this when mounting the secrets/configmaps into the build pod.  (generate our own volume names)

**Pull request merged**
